pkg/didcomm/transport/ws/outbound.go

```diff
@@ -15,57 +15,95 @@ import (
 	"nhooyr.io/websocket"
 
 	"github.com/hyperledger/aries-framework-go/pkg/didcomm/common/service"
+	"github.com/hyperledger/aries-framework-go/pkg/didcomm/protocol/decorator"
+	"github.com/hyperledger/aries-framework-go/pkg/didcomm/transport"
 )
 
 const webSocketScheme = "ws"
 
 // OutboundClient websocket outbound.
 type OutboundClient struct {
+	pool *connPool
+	prov transport.Provider
 }
 
 // NewOutbound creates a client for Outbound WS transport.
 func NewOutbound() *OutboundClient {
 	return &OutboundClient{}
 }
 
+// Start starts the outbound transport.
+func (cs *OutboundClient) Start(prov transport.Provider) error {
+	cs.pool = getConnPool(prov)
+	cs.prov = prov
+
+	return nil
+}
+
 // Send sends a2a data via WS.
 func (cs *OutboundClient) Send(data []byte, destination *service.Destination) (string, error) {
 	if destination.ServiceEndpoint == "" {
 		return "", errors.New("url is mandatory")
 	}
 
-	client, _, err := websocket.Dial(context.Background(), destination.ServiceEndpoint, nil)
-	if err != nil {
-		return "", fmt.Errorf("websocket client : %w", err)
-	}
-
-	defer func() {
-		err = client.Close(websocket.StatusNormalClosure, "closing the connection")
-		if err != nil && websocket.CloseStatus(err) != websocket.StatusNormalClosure {
-			logger.Errorf("failed to close connection: %v", err)
-		}
-	}()
-
-	ctx := context.Background()
+	conn, cleanup, err := cs.getConnection(destination)
+	defer cleanup()
 
-	err = client.Write(ctx, websocket.MessageText, data)
 	if err != nil {
-		return "", fmt.Errorf("websocket write message : %w", err)
+		return "", fmt.Errorf("get websocket connection : %w", err)
 	}
 
-	messageType, message, err := client.Read(ctx)
+	err = conn.Write(context.Background(), websocket.MessageText, data)
 	if err != nil {
-		return "", fmt.Errorf("websocket read message : %w", err)
-	}
-
-	if messageType != websocket.MessageText {
-		return "", errors.New("message type is not text message")
+		return "", fmt.Errorf("websocket write message : %w", err)
 	}
 
-	return string(message), nil
+	return "", nil
 }
 
 // Accept checks for the url scheme.
 func (cs *OutboundClient) Accept(url string) bool {
 	return strings.HasPrefix(url, webSocketScheme)
 }
+
+func (cs *OutboundClient) getConnection(destination *service.Destination) (*websocket.Conn, func(), error) {
+	if cs.pool != nil {
```

Comment:
Is this condition valid? Pool should have been set in start. Otherwise, you will have to put locks around it.

Reply:
Very valid point. Pool is already set in start(). This condition is not necessary and [removed](https://github.com/hyperledger/aries-framework-go/compare/5ba26394828508c6b430b55d5c3960891ca1dc81..89537285cd7d33cf2940c005e4a583162a90d55b#diff-84ebb4df3b345595955c03cc5f620fd9L70) in new patch.